File: bench/protocol_compare.py
#!/usr/bin/env python3
"""Apples-to-apples comparison of two streaming protocols for Gondolin:

  JSON+base64 : {"id": N, "stream": {"kind": "stdout", "b64": "<base64>"}}
  msgpack-rpc : msgpack-packed equivalent with `data` as msgpack bin (raw bytes)

Strategy:
  - Generate representative chunk corpora (binary 256KB/4MB, text 1MB, line-oriented 64KB).
  - For each corpus, simulate the daemon→wrapper pipeline:
      ENCODE side  : take chunk bytes, wrap in protocol envelope, serialize, framing.
      DECODE side  : read frame, deserialize, extract bytes payload.
  - Time encode+decode end-to-end. Repeat N iterations for stability.

We measure both implementations on the SAME chunk corpora, including raw
binary bytes (no UTF-8 corruption, because we're simulating the
post-fix daemon). The point is to isolate the protocol cost from the
Gondolin-side bug.
"""
from __future__ import annotations

import base64
import io
import json
import os
import statistics
import struct
import sys
import time

try:
    import msgpack  # type: ignore
except ImportError:
    sys.exit("FATAL: msgpack not installed")


# -------- Protocol A: JSON + base64 with newline framing --------

def encode_json_b64(chunks: list[bytes]) -> bytes:
    """Daemon side: turn a list of chunk bytes into a stream of newline-delimited JSON frames."""
    out = io.BytesIO()
    for i, chunk in enumerate(chunks):
        frame = {
            "id": 1,
            "stream": {
                "kind": "stdout",
                "b64": base64.b64encode(chunk).decode("ascii"),
            },
        }
        out.write(json.dumps(frame).encode("utf-8"))
        out.write(b"\n")
    # Terminal result frame
    out.write(json.dumps({"id": 1, "result": {"exit_code": 0}}).encode("utf-8"))
    out.write(b"\n")
    return out.getvalue()


def decode_json_b64(wire: bytes) -> tuple[list[bytes], int]:
    """Wrapper side: parse newline-delimited JSON frames, decode b64 data."""
    decoded: list[bytes] = []
    buf = bytearray(wire)
    exit_code = -1
    while buf:
        nl = buf.find(b"\n")
        if nl < 0:
            break
        line = bytes(buf[:nl])
        del buf[:nl + 1]
        frame = json.loads(line.decode("utf-8"))
        if "stream" in frame:
            decoded.append(base64.b64decode(frame["stream"]["b64"]))
        elif "result" in frame:
            exit_code = int(frame["result"].get("exit_code", -1))
            break
    return decoded, exit_code


# -------- Protocol B: msgpack with u32-length prefix framing --------

def encode_msgpack(chunks: list[bytes]) -> bytes:
    """Daemon side: turn chunks into length-prefixed msgpack frames.
    Frame: [u32 BE length][msgpack payload]
    """
    out = io.BytesIO()
    for chunk in chunks:
        frame = {
            "id": 1,
            "stream": {"kind": "stdout", "data": chunk},  # data is bytes -> msgpack bin
        }
        packed = msgpack.packb(frame, use_bin_type=True)
        out.write(struct.pack(">I", len(packed)))
        out.write(packed)
    final = msgpack.packb({"id": 1, "result": {"exit_code": 0}}, use_bin_type=True)
    out.write(struct.pack(">I", len(final)))
    out.write(final)
    return out.getvalue()


def decode_msgpack(wire: bytes) -> tuple[list[bytes], int]:
    decoded: list[bytes] = []
    pos = 0
    exit_code = -1
    while pos < len(wire):
        (n,) = struct.unpack_from(">I", wire, pos)
        pos += 4
        frame = msgpack.unpackb(wire[pos:pos + n], raw=False)
        pos += n
        if "stream" in frame:
            decoded.append(frame["stream"]["data"])  # already bytes
        elif "result" in frame:
            exit_code = int(frame["result"].get("exit_code", -1))
            break
    return decoded, exit_code


# -------- Corpora --------

def gen_binary(total: int, chunk_size: int = 15_000) -> list[bytes]:
    # Gondolin emits ~15KB chunks per the earlier bench. Replicate that.
    blob = os.urandom(total)
    return [blob[i:i + chunk_size] for i in range(0, len(blob), chunk_size)]


def gen_text_yes(total: int, chunk_size: int = 8_192) -> list[bytes]:
    # Pattern of `yes | head -c N` is "y\n" * (N/2). Realistic 8KB chunks.
    payload = (b"y\n" * (total // 2))[:total]
    return [payload[i:i + chunk_size] for i in range(0, len(payload), chunk_size)]


def gen_line_oriented(line_count: int = 1000) -> list[bytes]:
    # 1000 lines of "line-N\n" — average ~9 bytes; Gondolin chunks aggressively here.
    # From the real bench: 149 chunks for 1000 lines, mean ~60B per chunk.
    payload = b"".join(f"line-{i}\n".encode() for i in range(1, line_count + 1))
    chunk_size = max(1, len(payload) // 149)
    return [payload[i:i + chunk_size] for i in range(0, len(payload), chunk_size)]


def gen_small_echo() -> list[bytes]:
    return [b"hello-world\n"]


# -------- Driver --------

def bench(name: str, chunks: list[bytes], iters: int = 5) -> None:
    payload_bytes = sum(len(c) for c in chunks)
    n_chunks = len(chunks)
    print(f"\n=== {name} ===  (payload={payload_bytes:,} B in {n_chunks} chunks)")
    print(f"{'protocol':<14} {'wire B':>12} {'wire/pl':>8} {'enc_ms':>8} {'dec_ms':>8} {'total_ms':>9}")

    for label, enc, dec in [
        ("json+b64", encode_json_b64, decode_json_b64),
        ("msgpack", encode_msgpack, decode_msgpack),
    ]:
        encs = []
        decs = []
        wire_len = None
        for _ in range(iters):
            t0 = time.perf_counter()
            wire = enc(chunks)
            t1 = time.perf_counter()
            recovered, ec = dec(wire)
            t2 = time.perf_counter()
            encs.append((t1 - t0) * 1000)
            decs.append((t2 - t1) * 1000)
            wire_len = len(wire)
            # Correctness check
            joined = b"".join(recovered)
            joined_in = b"".join(chunks)
            if joined != joined_in:
                print(f"  !! {label} CORRUPTED ROUNDTRIP: got {len(joined)}B vs {len(joined_in)}B")
        enc_med = statistics.median(encs)
        dec_med = statistics.median(decs)
        ratio = wire_len / payload_bytes if payload_bytes else float("inf")
        print(f"{label:<14} {wire_len:>12,} {ratio:>8.3f} {enc_med:>8.2f} {dec_med:>8.2f} {enc_med+dec_med:>9.2f}")


def main():
    print(f"msgpack version: {msgpack.version}")
    print(f"python: {sys.version.split()[0]}")
    bench("01 small text echo", gen_small_echo(), iters=200)
    bench("02 256KB binary",   gen_binary(256 * 1024), iters=30)
    bench("03 4MB binary",     gen_binary(4 * 1024 * 1024), iters=10)
    bench("04 1MB text (yes)", gen_text_yes(1_000_000), iters=20)
    bench("05 64KB line-oriented", gen_line_oriented(1000), iters=200)


if __name__ == "__main__":
    main()

File: bench/protocol_compare.mjs
// Apples-to-apples node-side comparison of JSON+base64 vs msgpack
// for the gondolin streaming protocol.
//
// Run: node bench_protocol_compare.mjs
//
// Mirrors bench_protocol_compare.py on the Python side: feeds the same
// kind of chunk corpora through both encoders and reports wire size +
// encode time. (Decode-on-daemon isn't a real hot path — the daemon
// only encodes — but we time it anyway for completeness.)

import { performance } from "node:perf_hooks";
import { Buffer } from "node:buffer";
import * as msgpack from "@msgpack/msgpack";

// ---- Protocol A: JSON + base64 + newline framing ----

function encodeJsonB64(chunks) {
  const out = [];
  for (const chunk of chunks) {
    const frame = {
      id: 1,
      stream: { kind: "stdout", b64: chunk.toString("base64") },
    };
    out.push(Buffer.from(JSON.stringify(frame) + "\n", "utf8"));
  }
  out.push(Buffer.from(JSON.stringify({ id: 1, result: { exit_code: 0 } }) + "\n", "utf8"));
  return Buffer.concat(out);
}

function decodeJsonB64(wire) {
  const decoded = [];
  let buf = wire;
  let exitCode = -1;
  while (buf.length > 0) {
    const nl = buf.indexOf(0x0a);
    if (nl < 0) break;
    const line = buf.slice(0, nl).toString("utf8");
    buf = buf.slice(nl + 1);
    const frame = JSON.parse(line);
    if (frame.stream) {
      decoded.push(Buffer.from(frame.stream.b64, "base64"));
    } else if (frame.result) {
      exitCode = frame.result.exit_code | 0;
      break;
    }
  }
  return { decoded, exitCode };
}

// ---- Protocol B: msgpack with u32 BE length prefix ----

function encodeMsgpack(chunks) {
  const out = [];
  for (const chunk of chunks) {
    const frame = { id: 1, stream: { kind: "stdout", data: chunk } };
    const packed = msgpack.encode(frame);  // Uint8Array
    const len = Buffer.alloc(4);
    len.writeUInt32BE(packed.length, 0);
    out.push(len);
    out.push(Buffer.from(packed.buffer, packed.byteOffset, packed.byteLength));
  }
  const final = msgpack.encode({ id: 1, result: { exit_code: 0 } });
  const len = Buffer.alloc(4);
  len.writeUInt32BE(final.length, 0);
  out.push(len);
  out.push(Buffer.from(final.buffer, final.byteOffset, final.byteLength));
  return Buffer.concat(out);
}

function decodeMsgpack(wire) {
  const decoded = [];
  let pos = 0;
  let exitCode = -1;
  while (pos < wire.length) {
    const n = wire.readUInt32BE(pos);
    pos += 4;
    const frame = msgpack.decode(wire.slice(pos, pos + n));
    pos += n;
    if (frame.stream) {
      decoded.push(Buffer.from(frame.stream.data));
    } else if (frame.result) {
      exitCode = frame.result.exit_code | 0;
      break;
    }
  }
  return { decoded, exitCode };
}

// ---- Corpora ----

function randBytes(n) {
  const b = Buffer.allocUnsafe(n);
  for (let i = 0; i < n; i++) b[i] = (Math.random() * 256) | 0;
  return b;
}

function splitInto(buf, chunkSize) {
  const chunks = [];
  for (let i = 0; i < buf.length; i += chunkSize) chunks.push(buf.slice(i, i + chunkSize));
  return chunks;
}

function genBinary(total, chunkSize = 15_000) {
  return splitInto(randBytes(total), chunkSize);
}

function genTextYes(total, chunkSize = 8192) {
  const s = "y\n".repeat(Math.floor(total / 2));
  return splitInto(Buffer.from(s.slice(0, total), "utf8"), chunkSize);
}

function genLineOriented(lineCount = 1000) {
  const parts = [];
  for (let i = 1; i <= lineCount; i++) parts.push(`line-${i}\n`);
  const payload = Buffer.from(parts.join(""), "utf8");
  const chunkSize = Math.max(1, Math.floor(payload.length / 149));
  return splitInto(payload, chunkSize);
}

function genSmallEcho() {
  return [Buffer.from("hello-world\n", "utf8")];
}

// ---- Driver ----

function median(arr) {
  const s = [...arr].sort((a, b) => a - b);
  return s[Math.floor(s.length / 2)];
}

function bench(name, chunks, iters = 5) {
  const payloadBytes = chunks.reduce((s, c) => s + c.length, 0);
  console.log(`\n=== ${name} ===  (payload=${payloadBytes.toLocaleString()} B in ${chunks.length} chunks)`);
  console.log(
    `${"protocol".padEnd(14)} ${"wire B".padStart(12)} ${"wire/pl".padStart(8)} ${"enc_ms".padStart(8)} ${"dec_ms".padStart(8)} ${"total_ms".padStart(9)}`
  );

  for (const [label, enc, dec] of [
    ["json+b64", encodeJsonB64, decodeJsonB64],
    ["msgpack", encodeMsgpack, decodeMsgpack],
  ]) {
    const encs = [];
    const decs = [];
    let wireLen = 0;
    for (let i = 0; i < iters; i++) {
      const t0 = performance.now();
      const wire = enc(chunks);
      const t1 = performance.now();
      const { decoded } = dec(wire);
      const t2 = performance.now();
      encs.push(t1 - t0);
      decs.push(t2 - t1);
      wireLen = wire.length;
      // Correctness
      const joined = Buffer.concat(decoded);
      const joinedIn = Buffer.concat(chunks);
      if (!joined.equals(joinedIn)) {
        console.log(`  !! ${label} CORRUPTED ROUNDTRIP: got ${joined.length}B vs ${joinedIn.length}B`);
      }
    }
    const ratio = payloadBytes > 0 ? (wireLen / payloadBytes).toFixed(3) : "inf";
    const encMed = median(encs).toFixed(2);
    const decMed = median(decs).toFixed(2);
    const totMed = (Number(encMed) + Number(decMed)).toFixed(2);
    console.log(
      `${label.padEnd(14)} ${wireLen.toLocaleString().padStart(12)} ${String(ratio).padStart(8)} ${encMed.padStart(8)} ${decMed.padStart(8)} ${totMed.padStart(9)}`
    );
  }
}

bench("01 small text echo", genSmallEcho(), 500);
bench("02 256KB binary", genBinary(256 * 1024), 30);
bench("03 4MB binary", genBinary(4 * 1024 * 1024), 10);
bench("04 1MB text (yes)", genTextYes(1_000_000), 20);
bench("05 64KB line-oriented", genLineOriented(1000), 500);
